snappy/security.go

```diff
@@ -20,94 +20,241 @@
+	// SecurityOverride is a override for the high level security json
+	SecurityOverride *SecurityOverrideDefinition `yaml:"security-override,omitempty" json:"security-override,omitempty"`
+	// SecurityPolicy is a hand-crafted low-level policy
+	SecurityPolicy *SecurityPolicyDefinition `yaml:"security-policy,omitempty" json:"security-policy,omitempty"`
+
+	// SecurityCaps is are the apparmor/seccomp capabilities for an app
+	SecurityCaps []string `yaml:"caps,omitempty" json:"caps,omitempty"`
+}
+
+// securityPolicyType is a kind of securityPolicy, we currently
+// have "apparmor" and "seccomp"
+type securityPolicyType struct {
+	name          string
+	basePolicyDir string
+}
+
+var securityPolicyTypeAppArmor = securityPolicyType{
+	name:          "apparmor",
+	basePolicyDir: "/usr/share/apparmor/easyprof",
+}
+
+var securityPolicyTypeSeccomp = securityPolicyType{
+	name:          "seccomp",
+	basePolicyDir: "/usr/share/seccomp",
+}
+
+func (sp *securityPolicyType) policyDir() string {
+	return filepath.Join(dirs.GlobalRootDir, sp.basePolicyDir)
+}
+
+func (sp *securityPolicyType) frameworkPolicyDir() string {
+	frameworkPolicyDir := filepath.Join(policy.SecBase, sp.name)
+	return filepath.Join(dirs.GlobalRootDir, frameworkPolicyDir)
+}
 
-	// FIXME: this is snappy specific, on other systems like the
-	//        phone we may want different defaults.
-	if t.Template == "" && t.PolicyGroups == nil {
-		t.PolicyGroups = defaultPolicyGroups
+func (sp *securityPolicyType) findTemplate(template string) (string, error) {
+	if template == "" {
+		template = defaultTemplate
 	}
 
-	// never write a null value out into the json
-	if t.PolicyGroups == nil {
-		t.PolicyGroups = []string{}
+	systemTemplate := ""
+	fwTemplate := ""
+	subdir := filepath.Join("templates", defaultPolicyVendor(), defaultPolicyVersion())
+	systemTemplate = filepath.Join(sp.policyDir(), subdir, template)
+	fwTemplate = filepath.Join(sp.frameworkPolicyDir(), "templates", template)
+
+	// Always prefer system policy
+	fns := []string{systemTemplate, fwTemplate}
+	for _, fn := range fns {
+		content, err := ioutil.ReadFile(fn)
+		if err == nil {
+			return string(content), nil
+		}
 	}
 
-	if t.Template == "" {
-		t.Template = defaultTemplate
+	return "", &errPolicyNotFound{"template", sp, template}
+}
+
+func (sp *securityPolicyType) findCaps(caps []string, template string) ([]string, error) {
+	// XXX: this is snappy specific, on other systems like the phone we may
+	// want different defaults.
+	if template == "" && caps == nil {
+		caps = defaultPolicyGroups
+	} else if caps == nil {
+		caps = []string{}
 	}
 
-	outStr, err := json.MarshalIndent(t, "", "  ")
-	if err != nil {
-		return nil, err
+	subdir := filepath.Join("policygroups", defaultPolicyVendor(), defaultPolicyVersion())
+	parent := filepath.Join(sp.policyDir(), subdir)
+	fwParent := filepath.Join(sp.frameworkPolicyDir(), "policygroups")
+
+	// Nothing to find if caps is empty
+	if len(caps) == 0 {
```

Comment:
(trivial: the above might be moved to before the path-joining logic, since it renders it irrelevant.